cpp/src/arrow/compute/expression_test.cc

```diff
@@ -1616,6 +1616,82 @@ TEST(Expression, SimplifyWithComparisonAndNullableCaveat) {
+    Simplify{is_in(field_ref("i32"), int32(), "[]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(2)))
+        .Expect(false);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[null]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(2)))
+        .Expect(false);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(equal(field_ref("i32"), literal(7)))
+        .Expect(true);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(equal(field_ref("i32"), literal(6)))
+        .Expect(false);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(3)))
+        .Expect(is_in(field_ref("i32"), int32(), "[5,7,9]", null_matching_behavior));
+
+    Simplify{
+        is_in(field_ref("i32"), int32(), "[1,null,3,5,null,7,9]", null_matching_behavior)
+    }
+        .WithGuarantee(greater(field_ref("i32"), literal(3)))
+        .Expect(is_in(field_ref("i32"), int32(), "[5,7,9]", null_matching_behavior));
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(9)))
+        .Expect(false);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(less_equal(field_ref("i32"), literal(0)))
+        .Expect(false);
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(0)))
+        .ExpectUnchanged();
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(
+            or_(equal(field_ref("i32"), literal(3)), is_null(field_ref("i32"))))
+        .ExpectUnchanged();
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(
+            and_(less_equal(field_ref("i32"), literal(7)),
+                 greater(field_ref("i32"), literal(4))))
+        .Expect(is_in(field_ref("i32"), int32(), "[5,7]", null_matching_behavior));
+
+    Simplify{is_in(field_ref("u32"), int8(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("u32"), literal(3)))
+        .Expect(is_in(field_ref("u32"), int8(), "[5,7,9]", null_matching_behavior));
+
+    Simplify{is_in(field_ref("u32"), int64(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("u32"), literal(3)))
+        .Expect(is_in(field_ref("u32"), int64(), "[5,7,9]", null_matching_behavior));
+  }
+
+  Simplify{
+      is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", SetLookupOptions::INCONCLUSIVE)
+  }
+      .WithGuarantee(greater(field_ref("i32"), literal(3)))
+      .ExpectUnchanged();
+}
+
 TEST(Expression, SimplifyThenExecute) {
```

Comment:
I'd like to see more tests with null in the value_set, and some tests running against actual arrays. `TEST(Expression, SimplifyThenExecute)` should serve as a guide

Reply:
Added a test to actually execute against arrays. For value sets containing nulls, I left a comment about this above.